### PR TITLE
Do not concatenate identical dictionaries

### LINKED ISSUE #504: Do not copy dictionary values when they are the same in `concat`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
For dictionary arrays, the code for the `concat` simply copies the data from multiple places into a single contigous block: https://github.com/apache/arrow-rs/blob/master/arrow/src/compute/kernels/concat.rs#L55

For dictionary arrays, this results in potentially duplicating the dictionary values. For example:

Array1:  "A, C, A, B"
Dictionary: `{A, B, C}`
Values: `{0, 2, 0, 1}`

Array2:  "C, C, A"
Dictionary: `{A, B, C}`
Values: `{2, 2, 0}`

Note the dictionary is the same  between Array 1 and Array 2. Concatenating them together results in
Array3:  "A, C, A, B, C, C, A"
Dictionary: `{A, B, C, A, B, C}`
Values: `{0, 2, 0, 1, 2, 2, 0}`

Note the duplicated values in the dictionary as well as the fact the other values aren't even used 🤣 

In IOx it turns out that we concatenate rather large numbers of such arrays leading to *very* large duplicated dictionaries (read all about the gory details here): https://github.com/influxdata/influxdb_iox/issues/1822

**Describe the solution you'd like**
When the dictionary has the exact same contents (e.g. same pointer) there is no need to copy the dictionary

Here is a reproducer showing the actual issue:
```rust
    fn test_string_dictionary_repeated_dictionary() {
        let array : DictionaryArray<Int8Type> = vec!["a", "a", "b", "c"].into_iter().collect();
        let array_copy: DictionaryArray<Int8Type> = array.data().clone().into();

        // dictionary is "a", "b", "c"
        assert_eq!(array.values(), &(Arc::new(StringArray::from(vec!["a", "b", "c"])) as ArrayRef));
        assert_eq!(array.keys(), &Int8Array::from(vec![0, 0, 1, 2]));

        // concatenate it with itself
        let combined = concat(&[&array_copy as _, &array as _])
            .unwrap();
        let combined = combined
            .as_any()
            .downcast_ref::<DictionaryArray<Int8Type>>()
            .unwrap();

        // Since dictionares are the same, no need a copy -- it could be the same "a", "b", "c"
        assert_eq!(combined.values(), &(Arc::new(StringArray::from(vec!["a", "b", "c"])) as ArrayRef),
                   "Actual: {:#?}", combined);
        assert_eq!(combined.keys(), &Int8Array::from(vec![0, 0, 1, 2, 0, 0, 1, 2]));
    }
```

The actual output shows that the resulting dictionary is `"a", "b", "c", "a", "b", "c"` when it was literally the same pointer

```
---- compute::kernels::concat::tests::test_string_dictionary_repeated_dictionary stdout ----
thread 'compute::kernels::concat::tests::test_string_dictionary_repeated_dictionary' panicked at 'assertion failed: `(left == right)`
  left: `StringArray
[
  "a",
  "b",
  "c",
  "a",
  "b",
  "c",
]`,
 right: `StringArray
[
  "a",
  "b",
  "c",
]`: Actual: DictionaryArray {keys: PrimitiveArray<Int8>
[
  0,
  0,
  1,
  2,
  3,
  3,
  4,
  5,
] values: StringArray
[
  "a",
  "b",
  "c",
  "a",
  "b",
  "c",
]}
', arrow/src/compute/kernels/concat.rs:528:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

**Describe alternatives you've considered**
There are times when the dictionary may not be the exact same pointer, but contain mostly/all the same values which would also be useful to compact, but I will file that under a separate ticket as it will likely involve a tradeoff between CPU and memory

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #504

# Rationale for this change
 
See ticket

# What changes are included in this PR?

Adds `ArrayData::ptr_eq` and then uses this to elide dictionary concatenation in MutableArrayData

# Are there any user-facing changes?

This changes the behaviour of concat, to no longer concatenate the dictionaries if they are the same. Not sure if this counts

### CODE REVIEW

**Comment by jhorstmann** [comment]
Looks good to me. I initially thought `ptr_eq` for dictionary values could get away with only comparing the data buffer, but in theory the same buffer could be combined with different offsets or validity. So really need to compare all buffers and child data as int this PR.

File: arrow/src/array/data.rs [hunk 1]
```diff
@@ -1155,6 +1155,41 @@ impl ArrayData {
             Ok(())
         })
     }
+
+    /// Returns true if this `ArrayData` is equal to `other`, using pointer comparisons
```
Comment (alamb) [inline arrow/src/array/data.rs:1159]: 👍

File: arrow/src/array/data.rs [hunk 2]
```diff
@@ -1155,6 +1155,41 @@ impl ArrayData {
...
+    /// to determine buffer equality. This is cheaper than `PartialEq::eq` but may
+    /// return false negatives
```
Comment (alamb) [inline arrow/src/array/data.rs:1161]: in what case would this return a false negative (to the "are these two pointers the same" question)?

File: arrow/src/compute/kernels/concat.rs [hunk 3]
```diff
@@ -525,4 +525,44 @@ mod tests {
...
+            combined.keys(),
+            &Int8Array::from(vec![0, 0, 1, 2, 0, 0, 1, 2])
+        );
+
+        // Should have reused the dictionary
+        assert!(array.data().child_data()[0].ptr_eq(&combined.data().child_data()[0]));
+        assert!(
+            array_copy.data().child_data()[0].ptr_eq(&combined.data().child_data()[0])
+        );
+    }
```
Comment (alamb) [inline arrow/src/compute/kernels/concat.rs:567]: Can we also add a test of concatenating three dictionaries -- where 2 use the same dictionary and one is a different dictionary?

**Review by alamb — COMMENTED** [review]
Looks really nice -- thank you @tustvold

File: arrow/src/array/data.rs (hunk 2 above)
Comment (tustvold) [reply arrow/src/array/data.rs:1161]: It may say two things are not equal when logically they are?

Comment (tustvold) [reply arrow/src/array/data.rs]: I updated the wording

File: arrow/src/compute/kernels/concat.rs (hunk 3 above)
Comment (tustvold) [reply arrow/src/compute/kernels/concat.rs:573]: Added

File: arrow/src/array/data.rs [hunk 4]
```diff
@@ -1155,6 +1155,41 @@ impl ArrayData {
             Ok(())
         })
     }
+
+    /// Returns true if this `ArrayData` is equal to `other`, using pointer comparisons
+    /// to determine buffer equality. This is cheaper than `PartialEq::eq` but may
+    /// return false when the arrays are logically equal
```
Comment (alamb) [inline arrow/src/array/data.rs:1161]: 👍

**Review by alamb — APPROVED** [review]

**Pull request merged**